### PR TITLE
Fix for Codesmell check step1

### PR DESCRIPTION
**Pull request by haimeiguo**
Code Smell completed the following categories: 
- close resources for database
- close resources for stream
- equals()
- unmodifiable collection
- Random

### CODE REVIEW

File: example/hadoop/src/main/java/org/apache/iotdb/hadoop/tsfile/TsFileHelper.java [hunk 1]
```diff
@@ -132,7 +134,10 @@ public static void main(String[] args) throws FileNotFoundException, IOException
     }
     writeTsFile(filePath);
     TsFileSequenceReader reader = new TsFileSequenceReader(filePath);
-    logger.info("Get file meta data: {}", reader.readFileMetadata());
-    reader.close();
+    try {
+      logger.info("Get file meta data: {}", reader.readFileMetadata());
+    } finally {
+      reader.close();
+    }
```
Comment (qiaojialin) [inline example/hadoop/src/main/java/org/apache/iotdb/hadoop/tsfile/TsFileHelper.java:141]: usually sonar recommand to put the closable resource into try()

E.g.

try (TsFileSequenceReader reader = new TsFileSequenceReader(filePath)) {
      logger.info("Get file meta data: {}", reader.readFileMetadata());
}

The resource in try() will be auto closed after run out of try(){}

**Review by qiaojialin — COMMENTED** [review]
Good work, there are so many code smell... Pay attention to the try() syntax.

File: flink-iotdb-connector/src/main/java/org/apache/iotdb/flink/Event.java [hunk 2]
```diff
@@ -32,8 +32,8 @@
   private List<TSDataType> types;
   private List<Object> values;
 
-  public Event(String device, Long timestamp, List<String> measurements, List<TSDataType> types,
-      List<Object> values) {
+  public Event(String device, final Long timestamp, final List<String> measurements, final List<TSDataType> types,
```
Comment (jixuan1989) [inline flink-iotdb-connector/src/main/java/org/apache/iotdb/flink/Event.java:35]: this `final` is strange... 
I do not know whether it takes any effect...

File: server/src/main/java/org/apache/iotdb/db/conf/directories/strategy/RandomOnDiskUsableSpaceStrategy.java [hunk 3]
```diff
@@ -18,15 +18,17 @@
...
 import java.util.ArrayList;
 import java.util.List;
-import java.util.Random;
 import org.apache.iotdb.db.exception.DiskSpaceInsufficientException;
 import org.apache.iotdb.db.utils.CommonUtils;
 
 public class RandomOnDiskUsableSpaceStrategy extends DirectoryStrategy {
 
-  private Random random = new Random(System.currentTimeMillis());
+  private SecureRandom random = new SecureRandom(Longs.toByteArray(System.currentTimeMillis()));
```
Comment (jixuan1989) [inline server/src/main/java/org/apache/iotdb/db/conf/directories/strategy/RandomOnDiskUsableSpaceStrategy.java:31]: how about `new SecureRandom(BytesUtils.longToBytes(System.currentTimeMillis(), 8));` 
Then you can avoid import com.google.common....

File: server/src/main/java/org/apache/iotdb/db/engine/merge/recover/MergeLogger.java [hunk 4]
```diff
@@ -46,6 +46,10 @@
 
   private BufferedWriter logStream;
 
+  public MergeLogger(MergeLogger another) {
```
Comment (jixuan1989) [inline server/src/main/java/org/apache/iotdb/db/engine/merge/recover/MergeLogger.java:49]: why add such an constructor?

File: server/src/test/java/org/apache/iotdb/db/conf/adapter/HyperLogLogTest.java [hunk 5]
```diff
@@ -36,7 +37,7 @@
   public void testStreamLibHll() {
     final int seed = 12345;
     // data on which to calculate distinct count
-    Random random = new Random(seed);
+    SecureRandom random = new SecureRandom(Ints.toByteArray(seed));
```
Comment (jixuan1989) [inline server/src/test/java/org/apache/iotdb/db/conf/adapter/HyperLogLogTest.java:40]: The same, try to use BytesUtils

File: server/src/test/java/org/apache/iotdb/db/engine/merge/MergeLogTest.java [hunk 6]
```diff
@@ -39,6 +39,8 @@
 import org.junit.After;
 import org.junit.Before;
 import org.junit.Test;
+import org.slf4j.Logger;
```
Comment (jixuan1989) [inline server/src/test/java/org/apache/iotdb/db/engine/merge/MergeLogTest.java:42]: why add these?

File: server/src/test/java/org/apache/iotdb/db/engine/merge/MergeTaskTest.java [hunk 7]
```diff
@@ -44,6 +44,8 @@
 import org.junit.After;
 import org.junit.Before;
 import org.junit.Test;
+import org.slf4j.Logger;
```
Comment (jixuan1989) [inline server/src/test/java/org/apache/iotdb/db/engine/merge/MergeTaskTest.java:47]: why add these?

File: server/src/test/java/org/apache/iotdb/db/integration/IOTDBGroupByIT.java [hunk 8]
```diff
@@ -28,6 +28,8 @@
 import org.junit.Test;
 
 import java.sql.*;
+import org.slf4j.Logger;
```
Comment (jixuan1989) [inline server/src/test/java/org/apache/iotdb/db/integration/IOTDBGroupByIT.java:31]: the same

File: server/src/test/java/org/apache/iotdb/db/integration/IOTDBGroupByInnerIntervalIT.java [hunk 9]
```diff
@@ -26,6 +26,8 @@
 import org.junit.Test;
 
 import java.sql.*;
+import org.slf4j.Logger;
```
Comment (jixuan1989) [inline server/src/test/java/org/apache/iotdb/db/integration/IOTDBGroupByInnerIntervalIT.java:29]: the same

File: server/src/test/java/org/apache/iotdb/db/integration/IoTDBAggregationAlignByDeviceIT.java [hunk 10]
```diff
@@ -32,6 +32,8 @@
 import org.junit.Assert;
 import org.junit.Before;
 import org.junit.Test;
+import org.slf4j.Logger;
```
Comment (jixuan1989) [inline server/src/test/java/org/apache/iotdb/db/integration/IoTDBAggregationAlignByDeviceIT.java:35]: the same

File: server/src/test/java/org/apache/iotdb/db/integration/IoTDBAggregationDeleteIT.java [hunk 11]
```diff
@@ -26,6 +26,8 @@
 import org.junit.Test;
 
 import java.sql.*;
+import org.slf4j.Logger;
```
Comment (jixuan1989) [inline server/src/test/java/org/apache/iotdb/db/integration/IoTDBAggregationDeleteIT.java:29]: the same

File: server/src/test/java/org/apache/iotdb/db/integration/IoTDBAggregationIT.java [hunk 12]
```diff
@@ -43,6 +43,8 @@
 import org.junit.Assert;
 import org.junit.Before;
 import org.junit.Test;
+import org.slf4j.Logger;
```
Comment (jixuan1989) [inline server/src/test/java/org/apache/iotdb/db/integration/IoTDBAggregationIT.java:46]: the same

File: server/src/test/java/org/apache/iotdb/db/integration/IoTDBAggregationLargeDataIT.java [hunk 13]
```diff
@@ -42,6 +42,8 @@
 import org.junit.Assert;
 import org.junit.Before;
 import org.junit.Test;
+import org.slf4j.Logger;
```
Comment (jixuan1989) [inline server/src/test/java/org/apache/iotdb/db/integration/IoTDBAggregationLargeDataIT.java:45]: the same

File: server/src/test/java/org/apache/iotdb/db/integration/IoTDBAggregationSmallDataIT.java [hunk 14]
```diff
@@ -42,6 +42,8 @@
 import org.junit.Assert;
 import org.junit.Before;
 import org.junit.Test;
+import org.slf4j.Logger;
```
Comment (jixuan1989) [inline server/src/test/java/org/apache/iotdb/db/integration/IoTDBAggregationSmallDataIT.java:45]: ?

File: server/src/test/java/org/apache/iotdb/db/integration/IoTDBAliasIT.java [hunk 15]
```diff
@@ -33,6 +33,8 @@
 import org.junit.Assert;
 import org.junit.BeforeClass;
 import org.junit.Test;
+import org.slf4j.Logger;
```
Comment (jixuan1989) [inline server/src/test/java/org/apache/iotdb/db/integration/IoTDBAliasIT.java:36]: ?

File: server/src/test/java/org/apache/iotdb/db/integration/IoTDBAlignByDeviceIT.java [hunk 16]
```diff
@@ -32,6 +32,8 @@
 import org.junit.Assert;
 import org.junit.BeforeClass;
 import org.junit.Test;
+import org.slf4j.Logger;
```
Comment (jixuan1989) [inline server/src/test/java/org/apache/iotdb/db/integration/IoTDBAlignByDeviceIT.java:35]: ?

File: server/src/test/java/org/apache/iotdb/db/integration/IoTDBAutoCreateSchemaIT.java [hunk 17]
```diff
@@ -35,6 +35,8 @@
 import java.sql.ResultSet;
 import java.sql.ResultSetMetaData;
 import java.sql.Statement;
+import org.slf4j.Logger;
```
Comment (jixuan1989) [inline server/src/test/java/org/apache/iotdb/db/integration/IoTDBAutoCreateSchemaIT.java:38]: ?

File: server/src/test/java/org/apache/iotdb/db/integration/IoTDBClearCacheIT.java [hunk 18]
```diff
@@ -36,6 +36,8 @@
 import org.junit.Assert;
 import org.junit.BeforeClass;
 import org.junit.Test;
+import org.slf4j.Logger;
```
Comment (jixuan1989) [inline server/src/test/java/org/apache/iotdb/db/integration/IoTDBClearCacheIT.java:39]: ?

File: server/src/test/java/org/apache/iotdb/db/integration/IoTDBFlushQueryMergeIT.java [hunk 19]
```diff
@@ -93,11 +95,14 @@ public void selectAllSQLTest() throws ClassNotFoundException {
       boolean hasResultSet = statement.execute("SELECT * FROM root");
       Assert.assertTrue(hasResultSet);
 
-      try (ResultSet resultSet = statement.getResultSet()) {
+      ResultSet resultSet = statement.getResultSet();
```
Comment (jixuan1989) [inline server/src/test/java/org/apache/iotdb/db/integration/IoTDBFlushQueryMergeIT.java:98]: why change to this?

File: server/src/test/java/org/apache/iotdb/db/sync/receiver/load/FileLoaderTest.java [hunk 20]
```diff
@@ -22,15 +22,16 @@
 import static org.junit.Assert.assertFalse;
 import static org.junit.Assert.assertTrue;
 
+import com.google.common.primitives.Ints;
```
Comment (jixuan1989) [inline server/src/test/java/org/apache/iotdb/db/sync/receiver/load/FileLoaderTest.java:25]: use BytesUtils instead

File: server/src/test/java/org/apache/iotdb/db/sync/receiver/recover/SyncReceiverLogAnalyzerTest.java [hunk 21]
```diff
@@ -22,14 +22,15 @@
 import static org.junit.Assert.assertFalse;
 import static org.junit.Assert.assertTrue;
 
+import com.google.common.primitives.Ints;
```
Comment (jixuan1989) [inline server/src/test/java/org/apache/iotdb/db/sync/receiver/recover/SyncReceiverLogAnalyzerTest.java:25]: the same

**Review by jixuan1989 — CHANGES_REQUESTED** [review]
Thanks for your hard work!
I pinned some problems but not all of them.

1. Why you import slf4j in many java files even though they do not use slf4j?
2.  Try to avoid use Longs and Ints in com.google.common. Instead, use BytesUtils.
3. I do not know why we have to add `final` in the head of each parameter for each function.
Please show the detailed reason.
If no, you'd better revert the changes, as it may cause too many conflicts.

File: example/hadoop/src/main/java/org/apache/iotdb/hadoop/tsfile/TsFileHelper.java (hunk 1 above)
Comment (wshao08) [reply example/hadoop/src/main/java/org/apache/iotdb/hadoop/tsfile/TsFileHelper.java]: Fixed

File: flink-iotdb-connector/src/main/java/org/apache/iotdb/flink/Event.java (hunk 2 above)
Comment (haimeiguo) [reply flink-iotdb-connector/src/main/java/org/apache/iotdb/flink/Event.java:35]: the code smell doc recommended we should not return or store modifiable object. So I added final keyword to make the object unmodifiable. Following is the detail: 

2.4.13 分类名称：不应该存储或直接返回可变成员.
可变对象是那些状态可以更改的对象。例如,数组是可变的,但不是字符串。不应该将可变类成员返回给调用者或直接接受和存储。这样做会使您容易受到类状态中意外更改的影响。
而是使用不可修改的“集合”(通过“集合”)。unmodifiableCollection”、“集合。或复制可变对象，并存储或返回该副本。
此规则检查数组、集合和日期是否没有直接存储或返回

Comment (jixuan1989) [reply flink-iotdb-connector/src/main/java/org/apache/iotdb/flink/Event.java:35]: but adding a "final" keyword take nothing effect..

Comment (haimeiguo) [reply flink-iotdb-connector/src/main/java/org/apache/iotdb/flink/Event.java]: you are right. we just reverted this commit and deleted 'final' keyword. Thank you!!

File: spark-tsfile/src/main/java/org/apache/iotdb/spark/tsfile/qp/optimizer/PhysicalOptimizer.java [hunk 22]
```diff
@@ -45,7 +45,6 @@ public PhysicalOptimizer(List<String> columnNames) {
     this.columnNames = columnNames;
   }
 
-  @SuppressWarnings("squid:S3776") // Suppress high Cognitive Complexity warning
```
[the remark this answers is not in the document]
Comment (wshao08) [reply spark-tsfile/src/main/java/org/apache/iotdb/spark/tsfile/qp/optimizer/PhysicalOptimizer.java]: Fixed

File: server/src/test/java/org/apache/iotdb/db/sync/receiver/recover/SyncReceiverLogAnalyzerTest.java (hunk 21 above)
Comment (wshao08) [reply server/src/test/java/org/apache/iotdb/db/sync/receiver/recover/SyncReceiverLogAnalyzerTest.java]: Fixed

File: server/src/test/java/org/apache/iotdb/db/sync/receiver/load/FileLoaderTest.java (hunk 20 above)
Comment (wshao08) [reply server/src/test/java/org/apache/iotdb/db/sync/receiver/load/FileLoaderTest.java]: Fixed

File: server/src/test/java/org/apache/iotdb/db/integration/IoTDBFlushQueryMergeIT.java (hunk 19 above)
Comment (wshao08) [reply server/src/test/java/org/apache/iotdb/db/integration/IoTDBFlushQueryMergeIT.java]: Fixed

File: server/src/test/java/org/apache/iotdb/db/integration/IoTDBClearCacheIT.java (hunk 18 above)
Comment (wshao08) [reply server/src/test/java/org/apache/iotdb/db/integration/IoTDBClearCacheIT.java]: Fixed

File: server/src/main/java/org/apache/iotdb/db/conf/directories/strategy/RandomOnDiskUsableSpaceStrategy.java (hunk 3 above)
Comment (wshao08) [reply server/src/main/java/org/apache/iotdb/db/conf/directories/strategy/RandomOnDiskUsableSpaceStrategy.java]: Use BytesUtils instead

File: server/src/test/java/org/apache/iotdb/db/conf/adapter/HyperLogLogTest.java (hunk 5 above)
Comment (wshao08) [reply server/src/test/java/org/apache/iotdb/db/conf/adapter/HyperLogLogTest.java]: Fixed

File: server/src/test/java/org/apache/iotdb/db/engine/merge/MergeLogTest.java (hunk 6 above)
Comment (wshao08) [reply server/src/test/java/org/apache/iotdb/db/engine/merge/MergeLogTest.java]: Removed

File: server/src/test/java/org/apache/iotdb/db/integration/IoTDBAutoCreateSchemaIT.java (hunk 17 above)
Comment (wshao08) [reply server/src/test/java/org/apache/iotdb/db/integration/IoTDBAutoCreateSchemaIT.java]: Removed

File: server/src/main/java/org/apache/iotdb/db/engine/merge/recover/MergeLogger.java (hunk 4 above)
Comment (wshao08) [reply server/src/main/java/org/apache/iotdb/db/engine/merge/recover/MergeLogger.java]: Removed

**Comment by jixuan1989** [comment]
Can this PR be maintained anymore? 
if not, let's close it.

**Pull request closed without merging**